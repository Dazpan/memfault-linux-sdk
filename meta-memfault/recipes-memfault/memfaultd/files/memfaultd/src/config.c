//! @file
//!
//! Copyright (c) Memfault, Inc.
//! See License.txt for details
//!
//! @brief
//! config init and handling implementation

#include "config.h"

#include <errno.h>
#include <fcntl.h>
#include <json-c/json.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include "builtin_conf.h"
#include "memfaultd.h"

struct MemfaultdConfig {
  sMemfaultd *memfaultd;
  json_object *base;
  json_object *runtime;
};

/**
 * @brief Write runtime tree to file
 *
 * @param handle
 */
static void prv_config_write_config(sMemfaultdConfig *handle) {
  char *runtime_path = memfaultd_generate_rw_filename(handle->memfaultd, "runtime.conf");
  if (!runtime_path) {
    return;
  }

  if (json_object_to_file_ext(runtime_path, handle->runtime,
                              JSON_C_TO_STRING_SPACED | JSON_C_TO_STRING_PRETTY) == -1) {
    fprintf(stderr, "config:: Failed to update runtime config file '%s'.\n", runtime_path);
  }
  free(runtime_path);
}

/**
 * @brief
 *
 * @param object Source object to merge from
 * @param objects Array of objects to merge into
 * @param len Length or array returned
 */
static void prv_config_merge_object_into_array(json_object *object,
                                               sMemfaultdConfigObject **objects, int *len) {
  if (!object) {
    return;
  }

  json_object_object_foreach(object, key, value) {
    int i = 0;
    bool matched = false;
    for (i = 0; i < (*len); ++i) {
      if (strcmp((*objects)[i].key, key) == 0) {
        matched = true;
        break;
      }
    }
    if (!matched) {
      ++(*len);
      *objects = realloc(*objects, sizeof(sMemfaultdConfigObject) * (*len));
      (*objects)[i].key = key;
    }

    switch (json_object_get_type(value)) {
      case json_type_boolean:
        (*objects)[i].type = kMemfaultdConfigTypeBoolean;
        (*objects)[i].value.b = json_object_get_boolean(value);
        break;
      case json_type_int:
        (*objects)[i].type = kMemfaultdConfigTypeInteger;
        (*objects)[i].value.d = json_object_get_int(value);
        break;
      case json_type_string:
        (*objects)[i].type = kMemfaultdConfigTypeString;
        (*objects)[i].value.s = json_object_get_string(value);
        break;
      case json_type_object:
        (*objects)[i].type = kMemfaultdConfigTypeObject;
        break;
      default:
        (*objects)[i].type = kMemfaultdConfigTypeUnknown;
        break;
    }
  }
}

/**
 * @brief Find object by parent&key in given tree
 *
 * @param tree Base object to find sub-object in
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @return json_object* Found object
 */
static json_object *prv_config_find_object(json_object *tree, const char *parent_key,
                                           const char *key) {
  if (!tree) {
    return NULL;
  }
  json_object *object = NULL;
  if (parent_key && strlen(parent_key) != 0) {
    if (!json_object_object_get_ex(tree, parent_key, &object) ||
        json_object_get_type(object) != json_type_object) {
      return NULL;
    }
  } else {
    object = tree;
  }
  if (object) {
    if (key && strlen(key) != 0) {
      if (json_object_object_get_ex(object, key, &object)) {
        return object;
      }
    } else {
      return object;
    }
  }
  return NULL;
}

/**
 * @brief Set object into parent&key in runtime tree
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 */
static void prv_config_set_object(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  json_object *val) {
  json_object *object;
  if (parent_key && strlen(parent_key)) {
    if (!json_object_object_get_ex(handle->runtime, parent_key, &object) ||
        json_object_get_type(object) != json_type_object) {
      object = json_object_new_object();
      json_object_object_add(handle->runtime, parent_key, object);
    }
  } else {
    object = handle->runtime;
  }
  json_object_object_add(object, key, val);

  prv_config_write_config(handle);
}

/**
 * @brief Deep copy of Source object into Destination
 *
 * @param a Destination
 * @param b Source
 */
static void prv_config_merge_objects(json_object *a, json_object *b) {
  json_object_object_foreach(b, key, val) {
    if (json_object_get_type(val) == json_type_object) {
      json_object *tmp;
      if (json_object_object_get_ex(a, key, &tmp) &&
          json_object_get_type(tmp) == json_type_object) {
        prv_config_merge_objects(tmp, val);
      } else {
        json_object_object_add(a, key, val);
        json_object_get(val);
      }
    } else {
      json_object_object_add(a, key, val);
      json_object_get(val);
    }
  }
}

char *memfaultd_config_generate_rw_filename(sMemfaultdConfig *handle, const char *filename) {
  const char *data_dir;
  char *file = NULL;
  if (memfaultd_config_get_string(handle, "", "data_dir", &data_dir) && strlen(data_dir) != 0) {
    file = malloc(strlen(data_dir) + strlen(filename) + 1 + 1);
    if (!file) {
      return NULL;
    }
    strcpy(file, data_dir);
    strcat(file, "/");
    strcat(file, filename);
  }

  return file;
}

/**
 * @brief Initialise the config object
 *
 * @param file Master config filename
 * @return memfaultd_config_h* config object
 */
sMemfaultdConfig *memfaultd_config_init(sMemfaultd *memfaultd, const char *file) {
  sMemfaultdConfig *handle = calloc(sizeof(sMemfaultdConfig), 1);

  handle->memfaultd = memfaultd;

  handle->base = json_tokener_parse((char *)builtin_conf);

  int fd;
  if ((fd = open(file, O_RDONLY)) == -1) {
    if (errno == ENOENT) {
      fprintf(stderr, "config:: Configuration file '%s' not found.\n", file);
    } else {
      fprintf(stderr, "config:: Unable to open configuration file '%s', %s.\n", file,
              strerror(errno));
    }
  } else {
    json_object *object = json_object_from_fd(fd);
    if (!object) {
      fprintf(stderr, "config:: Unable to parse configuration file  '%s': %s\n",
              file, json_util_get_last_err());
    } else {
      prv_config_merge_objects(handle->base, object);
      json_object_put(object);
    }
    close(fd);
  }

  char *runtime_path = memfaultd_config_generate_rw_filename(handle, "runtime.conf");

  if (!runtime_path) {
    // No runtime config, warn but continue.
    fprintf(stderr, "config:: No runtime_config defined, settings will not persist.\n");
    handle->runtime = json_object_new_object();
  } else {
    if ((fd = open(runtime_path, O_RDONLY)) == -1) {
      if (errno != ENOENT) {
        // Missing file is not an error
        fprintf(stderr, "config:: Unable to open configuration file '%s', %s.\n", runtime_path,
                strerror(errno));
      }
      handle->runtime = json_object_new_object();
    } else {
      if (!(handle->runtime = json_object_from_fd(fd))) {
        fprintf(stderr, "config:: Unable to parse configuration file  '%s'\n", runtime_path);
        handle->runtime = json_object_new_object();
      }
      close(fd);
    }
    free(runtime_path);
  }

  return handle;
}

/**
 * @brief Destroy the config object
 *
 * @param handle config object
 */
void memfaultd_config_destroy(sMemfaultdConfig *handle) {
  if (handle) {
    if (handle->base) {
      json_object_put(handle->base);
    }
    if (handle->runtime) {
      json_object_put(handle->runtime);
    }
    free(handle);
  }
}

/**
 * @brief Dump both configuration trees to stdout
 *
 * @param handle config object
 */
void memfaultd_config_dump_config(sMemfaultdConfig *handle, const char *file) {
  printf("Base configuration (%s):\n", file);
  printf("  %s\n\n", json_object_to_json_string(handle->base));
  printf("Runtime configuration:\n");
  printf("  %s\n\n", json_object_to_json_string(handle->runtime));
}

/**
 * @brief Set string in config object
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @param val Value to set
 */
void memfaultd_config_set_string(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                 const char *val) {
  prv_config_set_object(handle, parent_key, key, json_object_new_string(val));
}

/**
 * @brief Set integer in config object
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @param val Value to set
 */
void memfaultd_config_set_integer(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  int val) {
  prv_config_set_object(handle, parent_key, key, json_object_new_int(val));
}

/**
 * @brief Set boolean flag in config object
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @param val Value to set
 */
void memfaultd_config_set_boolean(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  bool val) {
  prv_config_set_object(handle, parent_key, key, json_object_new_boolean(val));
}

/**
 * @brief Get string from config object
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @param val Value returned
 * @return true Successfully added string to config
 * @return false Failed to add string
 */
bool memfaultd_config_get_string(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                 const char **val) {
  json_object *object;

  if (!(object = prv_config_find_object(handle->runtime, parent_key, key)) &&
      !(object = prv_config_find_object(handle->base, parent_key, key))) {
    fprintf(stderr, "config:: Failed to find config object %s:%s \n", parent_key ? parent_key : "",
            key ? key : "");
    return false;
  }

  if (json_object_get_type(object) != json_type_string) {
    fprintf(stderr, "config:: Object is not of type %s %s:%s \n", "string",
            parent_key ? parent_key : "", key ? key : "");
    return false;
  }

  *val = json_object_get_string(object);
  return true;
}

/**
 * @brief Get integer from config object
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @param val Value returned
 * @return true Successfully added integer to config
 * @return false Failed to add integer
 */
bool memfaultd_config_get_integer(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  int *val) {
  json_object *object;

  if (!(object = prv_config_find_object(handle->runtime, parent_key, key)) &&
      !(object = prv_config_find_object(handle->base, parent_key, key))) {
    fprintf(stderr, "config:: Failed to find config object %s:%s \n", parent_key ? parent_key : "",
            key ? key : "");
    return false;
  }

  if (json_object_get_type(object) != json_type_int) {
    fprintf(stderr, "config:: Object is not of type %s %s:%s \n", "int",
            parent_key ? parent_key : "", key ? key : "");
    return false;
  }

  *val = json_object_get_int(object);
  return true;
}

/**
 * @brief Get boolean flag from config object
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param key Key name to set
 * @param val Value returned
 * @return true Successfully added boolean to config
 * @return false Failed to add boolean
 */
bool memfaultd_config_get_boolean(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  bool *val) {
  json_object *object;

  if (!(object = prv_config_find_object(handle->runtime, parent_key, key)) &&
      !(object = prv_config_find_object(handle->base, parent_key, key))) {
    fprintf(stderr, "config:: Failed to find config object %s:%s \n", parent_key ? parent_key : "",
            key ? key : "");
    return false;
  }

  if (json_object_get_type(object) != json_type_boolean) {
    fprintf(stderr, "config:: Object is not of type %s %s:%s \n", "boolean",
            parent_key ? parent_key : "", key ? key : "");
    return false;
  }

  *val = json_object_get_boolean(object);
  return true;
}

/**
 * @brief Get complete object structure
 *
 * @param handle config object
 * @param parent_key Parent key name, NULL for root object
 * @param objects Array of objects returned
 * @param len Length of array returned
 * @return true Successfully retrieved object structure
 * @return false Failed to retrieved structure
 */
bool memfaultd_config_get_objects(sMemfaultdConfig *handle, const char *parent_key,
                                  sMemfaultdConfigObject **objects, int *len) {
  *len = 0;
  *objects = NULL;

  json_object *object = NULL;
  if ((object = prv_config_find_object(handle->base, parent_key, "")) &&
      json_object_get_type(object) == json_type_object) {
    prv_config_merge_object_into_array(object, objects, len);
  }

  if ((object = prv_config_find_object(handle->runtime, parent_key, "")) &&
      json_object_get_type(object) == json_type_object) {
    prv_config_merge_object_into_array(object, objects, len);
  }

  return true;
}
