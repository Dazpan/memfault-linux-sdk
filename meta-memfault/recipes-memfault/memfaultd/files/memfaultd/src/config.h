//! @file
//!
//! Copyright (c) Memfault, Inc.
//! See License.txt for details
//!
//! @brief
//! config init and handling definition

#ifndef __MEMFAULT_config_H
#define __MEMFAULT_config_H

#include <memfaultd.h>
#include <stdbool.h>

typedef struct MemfaultdConfig sMemfaultdConfig;

sMemfaultdConfig *memfaultd_config_init(sMemfaultd *memfaultd, const char *file);
void memfaultd_config_destroy(sMemfaultdConfig *handle);
void memfaultd_config_dump_config(sMemfaultdConfig *handle, const char *file);

void memfaultd_config_set_string(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                 const char *val);
void memfaultd_config_set_integer(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  const int val);
void memfaultd_config_set_boolean(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  const bool val);

bool memfaultd_config_get_string(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                 const char **val);
bool memfaultd_config_get_integer(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  int *val);
bool memfaultd_config_get_boolean(sMemfaultdConfig *handle, const char *parent_key, const char *key,
                                  bool *val);
bool memfaultd_config_get_objects(sMemfaultdConfig *handle, const char *parent_key,
                                  sMemfaultdConfigObject **objects, int *len);
char *memfaultd_config_generate_rw_filename(sMemfaultdConfig *handle, const char *filename);

#endif
