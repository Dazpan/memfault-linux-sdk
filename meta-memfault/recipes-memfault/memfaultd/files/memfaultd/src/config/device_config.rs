//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use serde::{Deserialize, Serialize};

use crate::network::{DeviceConfigResponse, DeviceConfigResponseResolution, DeviceConfigRevision};

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
pub enum Resolution {
    Off,
    On,
}

impl From<DeviceConfigResponseResolution> for Resolution {
    fn from(resolution: DeviceConfigResponseResolution) -> Self {
        match resolution {
            DeviceConfigResponseResolution::Off => Resolution::Off,
            DeviceConfigResponseResolution::Low => Resolution::On,
            DeviceConfigResponseResolution::Normal => Resolution::On,
            DeviceConfigResponseResolution::High => Resolution::On,
        }
    }
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
pub struct Sampling {
    pub debugging_resolution: Resolution,
    pub logging_resolution: Resolution,
    pub monitoring_resolution: Resolution,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
/// DeviceConfig is configuration provided by Memfault backend.
pub struct DeviceConfig {
    pub revision: Option<DeviceConfigRevision>,
    pub sampling: Sampling,
}

impl From<DeviceConfigResponse> for DeviceConfig {
    fn from(response: DeviceConfigResponse) -> Self {
        Self {
            revision: Some(response.data.revision),
            sampling: Sampling {
                debugging_resolution: response
                    .data
                    .config
                    .memfault
                    .sampling
                    .debugging_resolution
                    .into(),
                logging_resolution: response
                    .data
                    .config
                    .memfault
                    .sampling
                    .logging_resolution
                    .into(),
                monitoring_resolution: response
                    .data
                    .config
                    .memfault
                    .sampling
                    .monitoring_resolution
                    .into(),
            },
        }
    }
}

impl Default for DeviceConfig {
    fn default() -> Self {
        Self {
            revision: None,
            sampling: Sampling {
                debugging_resolution: Resolution::Off,
                logging_resolution: Resolution::Off,
                monitoring_resolution: Resolution::Off,
            },
        }
    }
}

impl Sampling {
    pub fn development() -> Self {
        Self {
            debugging_resolution: Resolution::On,
            logging_resolution: Resolution::On,
            monitoring_resolution: Resolution::On,
        }
    }
}
