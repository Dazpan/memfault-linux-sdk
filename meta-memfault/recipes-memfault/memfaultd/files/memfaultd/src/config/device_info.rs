//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::fmt;
use std::process::Command;

#[derive(Debug)]
pub struct DeviceInfo {
    pub device_id: String,
    pub hardware_version: String,
}

#[derive(PartialEq, Eq, Debug)]
pub struct DeviceInfoWarning {
    line: String,
    message: &'static str,
}

impl std::fmt::Display for DeviceInfoWarning {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Skipped line: '{}' ({})", self.line, self.message)
    }
}

impl DeviceInfo {
    fn parse(output: &[u8]) -> eyre::Result<(DeviceInfo, Vec<DeviceInfoWarning>)> {
        let mut warnings = vec![];

        let mut di = DeviceInfo {
            device_id: String::new(),
            hardware_version: String::new(),
        };

        for line in std::str::from_utf8(output)?.lines() {
            if let Some((key, value)) = line.split_once('=') {
                match key {
                    "MEMFAULT_DEVICE_ID" => di.device_id = value.into(),
                    "MEMFAULT_HARDWARE_VERSION" => di.hardware_version = value.into(),
                    _ => warnings.push(DeviceInfoWarning {
                        line: line.into(),
                        message: "Unknown variable.",
                    }),
                }
            } else {
                warnings.push(DeviceInfoWarning {
                    line: line.into(),
                    message: "Expect '=' separated key/value pairs.",
                })
            }
        }
        match (
            di.device_id.is_empty(),
            di.hardware_version.is_empty(),
            !device_id_is_valid(&di.device_id),
        ) {
            (true, true, _) => Err(eyre::eyre!(
                "Missing both MEMFAULT_DEVICE_ID and MEMFAULT_HARDWARE_VERSION."
            )),
            (false, true, _) => Err(eyre::eyre!("Missing MEMFAULT_HARDWARE_VERSION.")),
            (true, false, _) => Err(eyre::eyre!("Missing MEMFAULT_DEVICE_ID.")),
            (_, _, true) => Err(eyre::eyre!("Invalid MEMFAULT_DEVICE_ID. Must be 1-128 characters long and contain only a-z, A-Z, 0-9, - and _")),
            (false, false, false) => Ok((di, warnings)),
        }
    }

    pub fn load() -> eyre::Result<(DeviceInfo, Vec<DeviceInfoWarning>)> {
        let user_output = Command::new("memfault-device-info").output()?;
        Self::parse(&user_output.stdout)
    }
}

fn device_id_is_valid(id: &str) -> bool {
    (1..=128).contains(&id.len())
        && id
            .chars()
            .all(|c| matches!(c, 'a'..='z' | 'A'..='Z' | '0'..='9' | '-' | '_'))
}

#[cfg(test)]
impl DeviceInfo {
    pub fn test_fixture() -> Self {
        DeviceInfo {
            device_id: "001".to_owned(),
            hardware_version: "DVT".to_owned(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[test]
    fn test_empty() {
        let r = DeviceInfo::parse(b"");
        assert!(r.is_err())
    }

    #[test]
    fn test_with_warnings() {
        let r =
            DeviceInfo::parse(b"MEMFAULT_DEVICE_ID=X\nMEMFAULT_HARDWARE_VERSION=Y\nblahblahblah\n");
        assert!(r.is_ok());

        let (di, warnings) = r.unwrap();
        assert_eq!(di.device_id, "X");
        assert_eq!(di.hardware_version, "Y");
        assert_eq!(warnings.len(), 1);
        assert_eq!(
            warnings[0],
            DeviceInfoWarning {
                line: "blahblahblah".into(),
                message: "Expect '=' separated key/value pairs."
            }
        );
    }

    #[rstest]
    // Minimum 1 character
    #[case("A", true)]
    // Allowed characters
    #[case(
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz_-",
        true
    )]
    // Disallowed characters
    #[case("DEMO.1234", false)]
    #[case("DEMO 1234", false)]
    // Too short (0 characters)
    #[case("", false)]
    // Too long (129 characters)
    #[case("012345679012345679012345679012345679012345679012345679012345679012345679012345679012345679012345679012345679012345678901234567890", false)]
    fn device_id_is_valid_works(#[case] device_id: &str, #[case] expected: bool) {
        assert_eq!(device_id_is_valid(device_id), expected);
    }
}
