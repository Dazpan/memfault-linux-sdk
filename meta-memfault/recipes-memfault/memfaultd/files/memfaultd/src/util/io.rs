//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::BufReader;

/// A trait for getting the length of a stream.
/// Note std::io::Seek also has a stream_len() method, but that method is fallible.
pub trait StreamLen {
    /// Gets the length of the stream in bytes.
    fn stream_len(&self) -> u64;
}

impl<R: StreamLen> StreamLen for BufReader<R> {
    fn stream_len(&self) -> u64 {
        self.get_ref().stream_len()
    }
}
