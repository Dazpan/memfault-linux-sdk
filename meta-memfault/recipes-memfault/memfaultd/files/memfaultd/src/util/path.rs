//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::ffi::OsStr;
use std::path::Path;

/// Splits the filename at the first dot.
/// This is similar to the nighly-only std::path::Path::file_prefix.
pub fn file_prefix(path: &Path) -> Option<&OsStr> {
    let file_name = path.file_name()?;
    file_name.to_str()?.split('.').next().map(OsStr::new)
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case("/test", "test")]
    #[case("/test.log", "test")]
    #[case("/test.log.zlib", "test")]
    fn test_file_prefix(#[case] path: &str, #[case] expected: &str) {
        assert_eq!(file_prefix(Path::new(path)), Some(OsStr::new(expected)));
    }
}
