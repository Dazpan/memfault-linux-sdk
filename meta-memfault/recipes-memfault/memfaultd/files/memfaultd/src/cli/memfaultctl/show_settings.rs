//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::borrow::Cow;
use std::io::{stdout, Write};
use std::path::Path;

use eyre::Result;

use crate::build_info::{BUILD_ID, GIT_COMMIT, VERSION};
use crate::config::{DeviceInfo, DeviceInfoWarning, JsonConfigs, MemfaultdConfig};

fn dump_config(
    writer: &mut impl Write,
    configs: &JsonConfigs,
    config_path: Option<&Path>,
) -> Result<()> {
    let path_str = config_path
        .map(Path::display)
        .map(|d| Cow::Owned(d.to_string()))
        .unwrap_or_else(|| Cow::Borrowed(MemfaultdConfig::DEFAULT_CONFIG_PATH));
    writeln!(writer, "Base configuration ({}):", path_str)?;
    writeln!(writer, "{}", serde_json::to_string_pretty(&configs.base)?)?;
    writeln!(writer)?;
    writeln!(writer, "Runtime configuration:")?;
    writeln!(
        writer,
        "{}",
        serde_json::to_string_pretty(&configs.runtime)?
    )?;
    Ok(())
}

type Device = (DeviceInfo, Vec<DeviceInfoWarning>);

fn dump_device_info(writer: &mut impl Write, device: &Device) -> Result<()> {
    let (device_info, warnings) = device;
    writeln!(writer, "Device configuration from memfault-device-info:")?;
    writeln!(writer, "  MEMFAULT_DEVICE_ID={}", device_info.device_id)?;
    writeln!(
        writer,
        "  MEMFAULT_HARDWARE_VERSION={}",
        device_info.hardware_version
    )?;

    if !warnings.is_empty() {
        writeln!(writer)?;
        for warning in warnings.iter() {
            writeln!(writer, "WARNING: {}", warning)?;
        }
    }

    Ok(())
}

struct Versions {
    version: &'static str,
    git_commit: &'static str,
    build_id: &'static str,
}

fn dump_version(writer: &mut impl Write, versions: &Versions) -> Result<()> {
    writeln!(writer, "Memfault version:")?;
    writeln!(writer, "  VERSION={}", versions.version)?;
    writeln!(writer, "  GIT COMMIT={}", versions.git_commit)?;
    writeln!(writer, "  BUILD ID={}", versions.build_id)?;
    Ok(())
}

fn dump_plugins(writer: &mut impl Write, plugins: &[&str]) -> Result<()> {
    writeln!(writer, "Plugin enabled:")?;
    for plugin in plugins {
        writeln!(writer, "  {}", plugin)?;
    }
    Ok(())
}

fn dump_settings(
    writer: &mut impl Write,
    configs: &JsonConfigs,
    config_path: Option<&Path>,
    device: &Device,
    versions: &Versions,
    plugins: &[&str],
) -> Result<()> {
    dump_config(writer, configs, config_path)?;
    writeln!(writer)?;
    dump_device_info(writer, device)?;
    writeln!(writer)?;
    dump_version(writer, versions)?;
    writeln!(writer)?;
    dump_plugins(writer, plugins)?;
    writeln!(writer)?;
    Ok(())
}

pub fn show_settings(config_path: Option<&Path>) -> Result<()> {
    let configs = MemfaultdConfig::parse_configs(config_path)?;
    let versions = Versions {
        version: VERSION,
        git_commit: GIT_COMMIT,
        build_id: BUILD_ID,
    };

    let enabled_plugins = [
        #[cfg(feature = "reboot")]
        "reboot",
        #[cfg(feature = "swupdate")]
        "swupdate",
        #[cfg(feature = "collectd")]
        "collectd",
        #[cfg(feature = "coredump")]
        "coredump",
        #[cfg(feature = "logging")]
        "logging",
    ];

    dump_settings(
        &mut stdout(),
        &configs,
        config_path,
        &DeviceInfo::load()?,
        &versions,
        &enabled_plugins,
    )
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;
    use std::path::PathBuf;

    use insta::assert_snapshot;
    use serde_json::json;

    use super::*;

    #[test]
    fn test() {
        let configs = JsonConfigs {
            base: json!({"project_key": "xyz"}),
            runtime: json!({"enable_data_collection": true}),
        };
        let config_path = PathBuf::from("/etc/memfaultd.conf");

        let device =
            DeviceInfo::parse(b"MEMFAULT_DEVICE_ID=X\nMEMFAULT_HARDWARE_VERSION=Y\nblahblahblah\n")
                .unwrap();

        let versions = Versions {
            version: "1.2.3",
            git_commit: "abcdef",
            build_id: "123456",
        };

        let enabled_plugins = ["reboot", "coredump"];

        let output = Vec::new();
        let mut writer = Cursor::new(output);
        dump_settings(
            &mut writer,
            &configs,
            Some(&config_path),
            &device,
            &versions,
            &enabled_plugins,
        )
        .unwrap();

        let output = String::from_utf8(writer.into_inner()).unwrap();
        assert_snapshot!(output);
    }
}
