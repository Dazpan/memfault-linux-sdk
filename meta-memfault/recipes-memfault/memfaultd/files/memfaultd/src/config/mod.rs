//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::{Path, PathBuf};

use crate::util::disk_size::DiskSize;

pub use self::{
    config_file::{JsonConfigs, MemfaultdConfig},
    device_info::{DeviceInfo, DeviceInfoWarning},
};
use eyre::Result;

mod config_file;
mod device_info;

/// Container of the entire memfaultd configuration.
/// Implement `From<Config>` trait to initialize module specific configuration (see `NetworkConfig` for example).
pub struct Config {
    pub device_info: DeviceInfo,
    pub config_file: MemfaultdConfig,
}

// See memfault-core-handler which places coredumps in this directory.
const COREDUMP_SUBDIRECTORY: &str = "core";
const LOGS_SUBDIRECTORY: &str = "logs";
const MAR_STAGING_SUBDIRECTORY: &str = "mar";

impl Config {
    pub fn read_from_system(user_config: Option<&Path>) -> Result<Self> {
        let config = MemfaultdConfig::load(user_config)?;

        let (device_info, warnings) = DeviceInfo::load()?;
        warnings.iter().for_each(|w| eprintln!("{}", w));

        Ok(Self {
            device_info,
            config_file: config,
        })
    }

    pub fn tmp_dir(&self) -> PathBuf {
        match self.config_file.tmp_dir {
            Some(ref tmp_dir) => tmp_dir.clone(),
            None => self.config_file.persist_dir.clone(),
        }
        .into()
    }

    pub fn tmp_dir_max_size(&self) -> DiskSize {
        DiskSize::new_capacity(self.config_file.tmp_dir_max_usage as u64)
    }

    pub fn tmp_dir_min_headroom(&self) -> DiskSize {
        DiskSize {
            bytes: self.config_file.tmp_dir_min_headroom as u64,
            inodes: self.config_file.tmp_dir_min_inodes as u64,
        }
    }

    pub fn coredumps_path(&self) -> PathBuf {
        self.tmp_dir().join(COREDUMP_SUBDIRECTORY)
    }

    pub fn logs_path(&self) -> PathBuf {
        self.tmp_dir().join(LOGS_SUBDIRECTORY)
    }

    pub fn mar_staging_path(&self) -> PathBuf {
        self.tmp_dir().join(MAR_STAGING_SUBDIRECTORY)
    }
}

#[cfg(test)]
impl Config {
    pub fn test_fixture() -> Self {
        Config {
            device_info: DeviceInfo::test_fixture(),
            config_file: MemfaultdConfig::test_fixture(),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use crate::{config::Config, util::path::AbsolutePath};

    #[test]
    fn tmp_dir_defaults_to_persist_dir() {
        let config = Config::test_fixture();

        assert_eq!(config.tmp_dir(), config.config_file.persist_dir);
    }

    #[test]
    fn tmp_folder_set() {
        let mut config = Config::test_fixture();
        let abs_path = PathBuf::from("/my/abs/path");
        config.config_file.tmp_dir = Some(AbsolutePath::try_from(abs_path.clone()).unwrap());

        assert_eq!(config.tmp_dir(), abs_path);
    }
}
