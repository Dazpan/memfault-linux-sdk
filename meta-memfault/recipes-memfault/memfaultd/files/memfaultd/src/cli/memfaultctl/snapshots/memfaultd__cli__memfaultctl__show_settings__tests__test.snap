---
source: memfaultd/src/cli/memfaultctl/show_settings.rs
expression: output
---
Base configuration (/etc/memfaultd.conf):
{
  "project_key": "xyz"
}

Runtime configuration:
{
  "enable_data_collection": true
}

Device configuration from memfault-device-info:
  MEMFAULT_DEVICE_ID=X
  MEMFAULT_HARDWARE_VERSION=Y

WARNING: Skipped line: 'blahblahblah' (Expect '=' separated key/value pairs.)

Memfault version:
  VERSION=1.2.3
  GIT COMMIT=abcdef
  BUILD ID=123456

Plugin enabled:
  reboot
  coredump
